### PR TITLE
Added an error if units provided are the wrong units

### LINKED ISSUE #666: Only strip expected units
Just an issue, not the work...

To avoid munging data inappropriately, include the expected units in the re used to remove units.  If it matches, strip the units and report a warning.  If it doesn't, issue an error.

> > I'm a bit skeptical of munging data like this, but I can see how it could be helpful if users add units frequently.
>
> IMHO, little issues add up.  And since the user is presented with a warning, it does get attention.  If you want, I could include the correct units in the re.  If they match: strip & warn.  If they don't: error.

### PR DESCRIPTION
**Pull request by hepcat72**
## Summary Change Description

During the review of a recent PR (#622), it was decided the stripping the units with only a warning was too risky, so I proposed in #666 to only strip "expected" units and otherwise issue an error that stops loading, so that's what this PR does.

While I was at it looking at units, I also added units to the Tracer Concentrations in the web and download templates.  They were already in the FCirc templates, but not the peak groups or peak data, which I believe was an oversight.  I'd gone to look at the website to get the expected tracer concentration units and was surprised they weren't there...

## Affected Issue Numbers

- Resolves #666

## Code Review Notes

Open to suggestions on strictness/flexibility and where to store the expected units that's not too "hard-coded" like it is now.

## Checklist

- [x] Changes have been added to the *Unreleased* section in the [`changelog.md`](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/changelog.md).
- [x] All issue requirements satisfied (or no linked issues)
- [x] [Linting passes](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#linting).
- [x] [Migrations created & committed *(or no model changes)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#migration-process)
- Tests
  - [x] [Tests implemented *(or no code changes)*](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#test-implementation)
  - [x] [All tests pass](https://github.com/Princeton-LSI-ResearchComputing/tracebase/blob/main/CONTRIBUTING.md#quality-control)
  - [x] All example load tests pass (remotely) *(or their failures predated and are unrelated to this branch)*

### CODE REVIEW

**Review by lparsons — CHANGES_REQUESTED** [review]
I like the additional functionality, though I don't think there is a need for a warning when expected units are found and stripped and the data is loaded.

**Comment by hepcat72** [comment]
> I like the additional functionality, though I don't think there is a need for a warning when expected units are found and stripped and the data is loaded.

I would have agreed with you had I not encountered a case where the warning revealed a problem.  In the cold exposure data, there was a subset of rows where units had been provided.  I think they were correct units, but Michael pointed out that another researcher had added them, but that the numbers were incorrect.

Units are not officially "allowed", which is why I originally made it a warning.  But I'm amenable to allowing "correct/supported" units and removing the warning.  We would have to update the docs WRT the documented format of the columns in the spreadsheet template.

File: DataRepo/tests/test_exceptions.py [hunk 1]
```diff
@@ -217,3 +219,45 @@ def test_summarize_int_list(self):
...
+        self.assertEqual(
+            (
+                "Units were stripped from values in 1 columns:\n"
+                "\tInfusion Rate (example: [3.3 ul/m/g] changed to: [3.3] on row(s): [5, 6])\n"
+                "Units are not allowed."
+            ),
+            str(una),
+        )
+        self.assertTrue(hasattr(una, "units_dict"))
+        self.assertEqual(una.units_dict, units_dict)
```
[the remark this answers is not in the document]
Comment (hepcat72) [reply DataRepo/tests/test_exceptions.py:263]: You don't think that modifying their entered value warrants a warning?  I would think that if you don't store exactly what they entered, you should let them know.

File: DataRepo/utils/exceptions.py [hunk 2]
```diff
@@ -166,17 +166,40 @@ class UnitsNotAllowed(Exception):
     def __init__(self, units_dict, message=None):
         if not message:
             nltab = "\n\t"
+            print(f"HEREHERE: {units_dict}")
```
[the remark this answers is not in the document]
Comment (hepcat72) [reply DataRepo/utils/exceptions.py:169]: - [ ] TODO: Remove debug print

File: DataRepo/utils/exceptions.py [hunk 3]
```diff
@@ -166,17 +166,40 @@ class UnitsNotAllowed(Exception):
...
+                    (
+                        f"{k} (example: [{units_dict[k]['example_val']}] changed to: "
+                        f"[{units_dict[k]['example_stripped']}] on row(s): {units_dict[k]['rows']})"
+                    )
+                    for k in units_dict.keys()
+                )
+            )
+            message = (
+                f"Units were stripped from values in {len(units_dict.keys())} columns:{nltab}{strip_str}\n"
+                "Units are not allowed."
```
[the remark this answers is not in the document]
Comment (hepcat72) [reply DataRepo/utils/exceptions.py:181]: I think a warning is warranted if you are changing what the user entered.  Say for example, they entered values using `minutes`.  They might try and search the database for a value in that field containing the substring "minutes".  I feel like we should warn them if we are going to change what they entered.

**Comment by hepcat72** [comment]
Re-requesting review @lparsons. I added some points of discussion to your blocking issue. Let me know if I've changed your mind.

File: DataRepo/utils/exceptions.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (lparsons) [reply DataRepo/utils/exceptions.py:181]: I'm OK with whichever way you decide to go, but I view this less like we are changing what they provided, but instead, interpreting what they provided. Similar to how we parse dates and then store it in our own format.

**Comment by lparsons** [comment]
In my opinion, we often interpret what they enter (e.g. dates, compounds, etc.) and store things as appropriate in the database, not as verbatim strings. If you'd like to leave it, that's fine, but let's try to make it clear that we did, in fact, store the value they provided assuming the provided unit. As it reads right now, it almost seems like the data wasn't valid and/or stored.

**Review by lparsons — APPROVED** [review]

File: DataRepo/utils/exceptions.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (hepcat72) [reply DataRepo/utils/exceptions.py:181]: That's fair.  You convinced me.

- [ ] TODO: No warning when expected units are stripped.

**Review by fkang-pu — APPROVED** [review]
LGTM!

**Pull request merged**